Họ tên: Nguyễn Minh Vũ

MSSV: 1712920

In [11]:
# # Run on Colab

# from google.colab import drive
# drive.mount('/content/drive')
# %cd /path/to/the/working/directory

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
/content/drive/My Drive/BT01


# BT01: Giới thiệu CUDA

## Câu 1

### 1.1 Biên dịch mã nguồn

In [3]:
!nvcc bt01_p1.cu -o bt01_p1

Cell biên dịch mã nguồn không báo lỗi. Như vậy chương trình đã được biên dịch thành công và được lưu thành file thực thi `bt01_p1`.

### 1.2 Thực thi chương trình với các `blockSize` khác nhau

#### a) `blockSize` 32 x 32 (mặc định)

In [4]:
!./bt01_p1 in.pnm out.pnm

Image size (width x height): 512 x 512

Processing time (use host): 2.535744 ms

GPU name: Tesla P100-PCIE-16GB
GPU compute capability: 6.0
Processing time (use device): 1.123680 ms

Error between device result and host result: 0.002674


Chương trình thực thi thành công. Ta thấy `device` chạy nhanh hơn `host` khoảng 2 lần. Độ lỗi cũng rất nhỏ, rơi vào 0.002674. So sánh với hướng dẫn của thầy thì con số này khá nhỏ (*0.xxx*) nên ta chấp nhận được.

#### b) `blockSize` 16 x 16

In [5]:
!./bt01_p1 in.pnm out.pnm 16 16

Image size (width x height): 512 x 512

Processing time (use host): 2.335264 ms

GPU name: Tesla P100-PCIE-16GB
GPU compute capability: 6.0
Processing time (use device): 1.092192 ms

Error between device result and host result: 0.002674


Chương trình thực thi thành công. Ta thấy `device` chạy nhanh hơn `host` khoảng 2 lần. Độ lỗi tương tự với `blockSize` 32x32 nên cũng chấp nhận được.

#### c) blockSize 64 x 64

In [6]:
!./bt01_p1 in.pnm out.pnm 64 64

Image size (width x height): 512 x 512

Processing time (use host): 2.470752 ms

GPU name: Tesla P100-PCIE-16GB
GPU compute capability: 6.0
Error: kernel did not execute properly
code: 9, reason: invalid configuration argument

Error between device result and host result: 119.969627


<a id='explain'></a>
Chương trình thực thi thành công, nhưng **báo lỗi khi chạy kernel**. Lỗi được trả về là `invalid configuration argument` - tham số cấu hình không hợp lệ. Việc chỉ định `blockSize` 64 x 64 nghĩa là ta yêu cầu `device` cung cấp các `block` có 64 x 64 = 4096 `thread`. Mà trong [bảng này](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#features-and-technical-specifications__technical-specifications-per-compute-capability), dòng **Maximum number of threads per block** (cho biết số thread tối đa trên một block) của tất cả *Compute Capability* hiện tại chỉ là 1024.

Do đó kernel không thể thực thi.

## Câu 2

### 2.1 Biên dịch mã nguồn

In [7]:
!nvcc bt01_p2.cu -o bt01_p2

Cell biên dịch mã nguồn không báo lỗi. Như vậy chương trình đã được biên dịch thành công và được lưu thành file thực thi `bt01_p2`.

### 2.2 Thực thi chương trình với các `blockSize` khác nhau

#### a) `blockSize` 32 x 32 (mặc định)

In [8]:
!./bt01_p2 in.pnm out2.pnm out2_target.pnm

Image size (width x height): 512 x 512

Processing time (use host): 274.851105 ms
Error: 0.000000

GPU name: Tesla P100-PCIE-16GB
GPU compute capability: 6.0
Processing time (use device): 1.419616 ms
Error: 0.000703



Chương trình thực thi thành công. Ta thấy `device` chạy nhanh hơn `host` gần 300 lần.

Độ lỗi có một chút khác biệt:
- khi chạy bằng `host`, độ lỗi bằng 0. Nghĩa là chương trình cho ra kết quả hoàn toàn khớp với ảnh test `out2_target.pnm`.
- khi chạy bằng `device`, độ lỗi nhích lên một chút, rơi vào 0.000703. So sánh với hướng dẫn của thầy thì con số này nhỏ hơn nhiều so với 0.xxx, nên đây là một kết quả chấp nhận được.

#### b) `blockSize` 16 x 16

In [9]:
!./bt01_p2 in.pnm out2.pnm out2_target.pnm 16 16

Image size (width x height): 512 x 512

Processing time (use host): 279.672546 ms
Error: 0.000000

GPU name: Tesla P100-PCIE-16GB
GPU compute capability: 6.0
Processing time (use device): 1.283040 ms
Error: 0.000703



Chương trình thực thi thành công. Ta thấy `device` chạy nhanh hơn `host` gần 300 lần. Độ lỗi tương tự với `blockSize` 32x32 nên chấp nhận được.

#### c) blockSize 64 x 64

In [10]:
!./bt01_p2 in.pnm out2.pnm out2_target.pnm 64 64

Image size (width x height): 512 x 512

Processing time (use host): 273.769287 ms
Error: 0.000000

GPU name: Tesla P100-PCIE-16GB
GPU compute capability: 6.0
Error: kernel did not run properly
Code: 9, Reason: invalid configuration argument

Error: 124.033791



Chương trình thực thi thành công, nhưng **báo lỗi khi chạy kernel**. Lỗi được trả về là `invalid configuration argument` - tham số cấu hình không hợp lệ. Ta lý giải tương tự mục [1.2c](#explain)